### DISCUSSION TITLE
Creating new aggregate and relevant event types in v1 and v2 to persist in EventStore

### DISCUSSION THREAD

**Discussion by vaishmanjunath** [zitadel/zitadel#8893] [Q&A] [answered]
I have cloned zitadel repository and currently have a branch created from tag v2.64.1.
I am trying to add an aggregate type say "abc", and some events under it like createabc, changeabc and removeabc.

I created a directory under /repository as /internal/repository/abc, and created below three files: (Similar to https://github.com/zitadel/zitadel/tree/v2.64.1/internal/repository/oidcsession)

- aggregate.go : Created a new aggregate type and version as v1, and added a function NewAggregate.
- abc.go: This defines some constants for abc related events, have struct definition for AbcAddedEvent (inhering evenstore.BaseEvent), then function that creates AbcAddedEvent based on input parameters.
- event_store.go: it adds above abc related events to registry. (eventstore.RegisterFilterEventMapper)

I have some questions here:
1. I have not defined any fields yet, want to test adding an entry into eventStore.events2 table. Do you recommend the above approach using v1? 

2. If yes, I was looking at eventStore definition in https://github.com/zitadel/zitadel/blob/v2.64.1/internal/eventstore/eventstore.go, was trying to find Pusher implementation. I see some implementations of Push: https://github.com/zitadel/zitadel/blob/v2.64.1/internal/eventstore/repository/sql/crdb.go and some in V3.
How to find the right Pusher implementation and create EventStore object in v1? (so that I can try persisting a record)

3. Which version of eventstore.go should I be using? I see v1 and v2, v2 comes with PushIntent and Reducer etc.
https://github.com/zitadel/zitadel/blob/v2.64.1/internal/v2/eventstore/event_store.go
https://github.com/zitadel/zitadel/blob/v2.64.1/internal/v2/eventstore/postgres/push.go
I believe the files I have created work only with v1 and not v2, and database has only events2 table. 

Can you please guide me on how to achieve this in v2? and point me to some examples on creating aggregate.go, abc.go and registering event (or something similar) in v2?

**Comment by hifabienne** [comment]
@adlerhurst and @muhlemmer  can you help here?

**Comment by muhlemmer** [comment]
In ZITADEL the `internal/commands` package is where we write events to the eventstore. You shouldn't need to interact with the underlying eventstore implementations directly.

Take this command for example:

https://github.com/zitadel/zitadel/blob/04a166f2d2392da59baf93e22c2f68223e986e7b/internal/command/web_key.go#L21-L49

1. We get some state with `c.getAllWebKeys`
2. We create commands (events which implement `eventstore.Command`) based on the business logic requirement, using the constructors from the `internal/repository/webkey` package
3. Commands are pushed using `c.pushAppendAndReduce` which uses `c.eventstore.Push` under the hood and then updates the model with the current state, required for the `CreateWebKey` return data:

https://github.com/zitadel/zitadel/blob/04a166f2d2392da59baf93e22c2f68223e986e7b/internal/command/command.go#L208-L217

In the `Commands` struct the `eventstore` field is this type:

https://github.com/zitadel/zitadel/blob/04a166f2d2392da59baf93e22c2f68223e986e7b/internal/eventstore/eventstore.go#L18-L31

The `Pusher`, `Querier` and `Searcher` interfaces are then implemented by different sub-packages (`v1`, `v3` etc).

>  Created a new aggregate type and version as v1

The aggregate version only applies to the aggregate / object you are managing and has nothing to do with the eventstore implementation. In theory we provided the aggregate versioning so you can change event definitions on the same aggregate later. However, support for this is not fully implemented and we are [considering removing](https://github.com/zitadel/zitadel/issues/5356) this property on events. I recently went wrong here myself and we had to [fix that](https://github.com/zitadel/zitadel/pull/8886).

If you're starting out to develop with ZITADEL, always use v1 aggregate versions. 

Happy hacking!

**Comment by vaishmanjunath** [comment]
@muhlemmer Thanks for the detailed response.
When I was talking about versions, I meant push implementation version as I came across v1, v2 and v3 push.go files.
Anyway now that I will have to use commands to write to the event store, it would not matter.

Some follow-up questions here:

1. If I need a simple push to the event store - events and fields table, below is command/abc.go - Does the below structure look fine?

`
        func (c *Commands) AddUserAbcData(ctx context.Context, userId string, abc string) (*domain.ObjectDetails, error) {

        // Create an aggregate with userId as aggregateId.

        // Generated an addEvent command (called addedCmd)

	commands := []eventstore.Command{addedCmd}

	model := NewUserAbcDataWriteModel(aggregate.ID, authz.GetInstance(ctx).InstanceID())

	err := c.pushAppendAndReduce(ctx, model, commands...)
	if err != nil {
		return nil, err
	}

	return writeModelToObjectDetails(&model.WriteModel), nil
}

        Writemodels: 
         type UserAbcDataWriteModel struct {
	eventstore.WriteModel
	UserID            string
	AbcID            string
         }

        func NewUserAbcDataWriteModel(aggregateID, resourceOwner string) *UserAbcDataWriteModel {
	return &UserAbcDataWriteModel{
		WriteModel: eventstore.WriteModel{
			AggregateID:   aggregateID,
			ResourceOwner: resourceOwner,
		},
	}
        }
`

Questions:
- Abc holds user relevant some data. Hence, planning to use userId as the aggregate Id, that should work right? It'll be easier to query

- Does this command.pushAppendAndReduce, write to both events2 and fields table? 

- The above stores only one Abc entry per user, how do I change command/abc.go functionality to handle adding multiple such Abc's per user in one shot? (some can be user.abc.added / user.abc.removed type of events), how would writeModel look like in that case? Can you please provide an example?

- In model.go I have only added AddUserAbcData function, not any reduce or append function definition. What are reduce and append functions that I see in https://github.com/zitadel/zitadel/blob/v2.64.1/internal/command/web_key_model.go?, 
I also noticed WebKeyWriteModels for multiple keys, I would need something similar - to create few events in bulk.

- Once multiple events are created in db, how do I query Abc field values of an user? can you please provide a reference? I would prefer results to be events with field values so that I can extract 'Abc' field values, and  I can maintain list for Abc strings to be derived from these fields, and based on event type: I can add or remove entries to the list.

- Events2 table will have **eventType: userEventAdded, aggregateType, aggregateId, sequence number**, and corresponding fields table will hold **aggType., aggId, objectId, version., and then field names and values**.  For an aggregateType + aggregateId + eventType row in events2 table, we just query all field table records, is it? How is the order maintained in query response?

-  **Most important:** Once I add command/abc.go, how can I test this locally? or through an rest endpoint (say for example)? Where can I find an existing command object that can be directly used?

Thank you!

**Reply by muhlemmer** [reply]
> Abc holds user relevant some data. Hence, planning to use userId as the aggregate Id, that should work right? It'll be easier to query

Sure, you can use the user ID. I just recommend against using the `user` aggregate as it is already a very large aggregate and it will become slow to query when too many events are in the same aggregate.


> Does this command.pushAppendAndReduce, write to both events2 and fields table?

No it does not. The fields table is a specialized case and the internal API isn't stable. It might even get dropped in the future. We are still evaluating if we want to keep it and how we should expose its internal API. I advise against using it for now.

> The above stores only one Abc entry per user, how do I change command/abc.go functionality to handle adding multiple such Abc's per user in one shot? (some can be user.abc.added / user.abc.removed type of events), how would writeModel look like in that case? Can you please provide an example?

You can push a slice of commands. So yes, you can push all mutations at once. You can look to other WriteModels in the command package for examples. WriteModels must implement the [`QueryReducer` interface](https://github.com/zitadel/zitadel/blob/ecbf0db15b6bd09587bdaebc2874ec7c207864ea/internal/eventstore/eventstore.go#L255-L259). The `AppendEvents` method queues the events inside `BaseWriteModel` and `Reduce` method applies the state changes from the queued events.

> In model.go I have only added AddUserAbcData function, not any reduce or append function definition. What are reduce and append functions that I see in https://github.com/zitadel/zitadel/blob/v2.64.1/internal/command/web_key_model.go?,
I also noticed WebKeyWriteModels for multiple keys, I would need something similar - to create few events in bulk.

As mentioned above, `Reduce` and `Append` functions apply events to a WriteModel. `WebKeyWriteModels` is there because we needed to query multiple aggregate IDs into a single model. Eg, all keys of an instance. Typically a WriteModel applies to a single aggregate ID.

Btw, WriteModels are only needed if you want to determine the current state of something, based on events. If you do not care about the state, you don't need a WriteModel.


> Once multiple events are created in db, how do I query Abc field values of an user? can you please provide a reference? I would prefer results to be events with field values so that I can extract 'Abc' field values, and I can maintain list for Abc strings to be derived from these fields, and based on event type: I can add or remove entries to the list.

Reading in zitadel happend in the `inernal/query` package. It is the counter part of the `command` package. You can go two ways:

1. If you know the aggregate ID, query the events and build a [`ReadModel`](https://github.com/zitadel/zitadel/blob/ecbf0db15b6bd09587bdaebc2874ec7c207864ea/internal/query/web_key.go#L29-L42).
2. If you need to query using other properties, like username you will need a `projection` table. Projections are updated by a spooler and are eventual consistent and optimized for repeated reads.
  - This is how the [`web_key` projection](https://github.com/zitadel/zitadel/blob/ecbf0db15b6bd09587bdaebc2874ec7c207864ea/internal/query/projection/web_key.go#L1-L163) is defined 
   - This how we [query](https://github.com/zitadel/zitadel/blob/ecbf0db15b6bd09587bdaebc2874ec7c207864ea/internal/query/web_key.go#L44-L68) the web key based on instance ID and active state

> Events2 table will have eventType: userEventAdded, aggregateType, aggregateId, sequence number, and corresponding fields table will hold aggType., aggId, objectId, version., and then field names and values. For an aggregateType + aggregateId + eventType row in events2 table, we just query all field table records, is it? How is the order maintained in query response?

The order is established by the internal query builder. There is also a `position` column for this. This is not something you need worry about as implementer.

> Most important: Once I add command/abc.go, how can I test this locally? or through an rest endpoint (say for example)? Where can I find an existing command object that can be directly used?

You can look at the various unit tests `internal/command`. We have helpers that create a mocked event store to which you can set expected calls.

There is also an [events API](https://zitadel.com/docs/apis/resources/admin/admin-service-list-events) you can use in production.

**Comment by vaishmanjunath** [comment]
@muhlemmer Thanks for the explanation.

- yes I have a different aggregate-type, its not user, so should be okay.

- If it does not write to both tables, where are the values stored? as payload in events table is it?

- If I do not need writeModels, I could loop over say input AbcIds, and create corresponding add/remove commands and directly invoke command.evenststore.Push, right?

-  If I do need writeModels to determine the final state of events after the push, all events received in push response, will be appended to BaseWriteModel Events using Append(), and the reduce() will iterate over BaseWriteModel events and have custom logic to store final state in WriteModel.

- About my question on testing, I went through mock unit tests, but I have postgres instance thats running and zitadel connects to it. I want to test userAbcData being actually stored in events table. How can I achieve this?

- When zitadel is up locally, UserServer is created (/internal/api/grpc/user/v2/server.go), I was thinking I could add a method to userServer, and call this method that adds entry to the events table. Does that sound feasible? I want to be able to hit REST endpoint to create this entry.

- If yes, then I need to register POST method, **pkg/grpc/user/v2/user_service.pb.gw.go** this has code to register handlers for each method, 
for example : sending email handler in <RegisterUserServiceHandlerServer> function is below. I could define a similar method here right? I think I will be able invoke this external endpoint. 
Is there anywhere else I need to make a change to register the endpoint?

`
         {
         
         mux.Handle("POST", pattern_UserService_SetEmail_0, func(w http.ResponseWriter, req *http.Request, pathParams 
         map[string]string) {
		// some code
		annotatedContext, err = runtime.AnnotateIncomingContext(ctx, mux, req, "/zitadel.user.v2.UserService/SetEmail", runtime.WithHTTPPathPattern("/v2/users/{user_id}/email"))
		// some code
		resp, md, err := local_request_UserService_SetEmail_0(annotatedContext, inboundMarshaler, server, req, pathParams)
		// some code

	})
`
I would need function similar to **local_request_UserService_SetEmail_0**  that ultimately invoke Commands method to create userAbcData.

- On another parallel lines, due to events table holding all the data which is ever growing. We cannot delete any entries as its the only source of truth. We are using self hosted version of zitadel. I am skeptical that events table being super large, will slow down the database and affect the performance. I plan to leverage the same postgres instance that is shared by few other services.
What would be resource requirements for database to handle large load?
Do you recommend any optimization techniques/ indexing? Should we worried about this at all?
Will horizontal scaling suffice? Some insights regarding cloud version of zitadel would help me understand performance aspect better.

**Reply by muhlemmer** [reply, marked as answer]
> If I do not need writeModels, I could loop over say input AbcIds, and create corresponding add/remove commands and directly invoke command.evenststore.Push, right?

That sounds like you do need state. Use a WriteModel.

Another example: The [Debug events API](https://github.com/zitadel/zitadel/tree/main/internal/api/grpc/resources/debug_events/debug_events) is the smallest implementation we have that can server as another good example how events are pushed and then projected.

**Regarding tests**:

Assuming this feature should end up in our repository, mock testing is the only way. There is no database running during the unit test. On the other hand, we do have integration tests which test the complete stack. This tests run a coverage server of ZITADEL and uses gPRC to make calls to the server and set/get state. I already gave a link to the events API that can be used. See also the [contribution guidelines](https://github.com/zitadel/zitadel/blob/main/CONTRIBUTING.md#run-local-unit-tests) on how to run integration tests.

**Regarding performance**:

ZITADEL already defines and creates the required indices on the eventstore and the size of the eventstore doesn't influence performance, as long as the size of a single aggregate ID is kept in check. We have literally billions of events in our Cloud server and have enterprise customer with even more. 


**Regarding this discussion**:

Whatever functionality or feature you are adding, please consider committing it back to us. You can start with a proposal in an issue and follow up with a PR. That way we will be able to look at the code and guide you better. If the intention is to develop the feature for private use, please be understanding we can't keep providing guidance at this level.

**Comment by vaishmanjunath** [comment]
Thank you, I totally understand! @muhlemmer 

I have one last topic, and your inputs will really help.

- I am trying to keep track of some related events in single aggregate, so I would need to use unique aggregateID. I am going to add a projection definition and define projection table insertion on different event types. (I have append only events, no removal event types) I would have the below structure in projection,

**aggregateId   |   userId  | created At  | sequence | AbcValue**

agg123           |  u1.        | t1                  | 1               | q1
agg123           |  u1.        | t1                  | 2              | q2

agg125           |  u1.        | t2                 | 1               | q3
agg125           |  u1.        | t2                 | 2               | q4

- I only need to get all latest values of **Abc** for a specific user (that would be q3, q4 here), and I would not know aggregateId, I believe I can get that with latest createdAt timestamp, and find the aggregate_id. (agg125), and then query rows where agg_id = agg125. With indexes on agg_id and user_id - query will be faster. 

- As the only relevant values for user here are with agg_id = agg125, I would need a way to get rid of older rows(q1, q2 valued rows) else even projection table would be ever growing like events-table. Do you recommend deleting older rows, and then inserting new rows in single transaction?  (can you point me to some examples where both projection queries : insert row and delete row are run together). or what's the best way to go about this and have only recent entries?, so that queries are simple.

- Is there any zitadel example like this? of append-only event types, and removal has to be handled separately in projections.

- Also, how do I store references (foreign key) in event store/projection tables? Can you provide a reference? Like userId in payload for totally different aggregate_type

**Comment by vaishmanjunath** [comment]
@muhlemmer 
Can I please get your inputs here? Thanks!
